File: src/step1_read_print.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <readline/readline.h>
#include <readline/history.h>

char* READ(char* token);
char* EVAL(char* expre);
char* PRINT(char* result);

int main(void) {

  rl_bind_key('\t', rl_complete);
  using_history();

  while (true) {

    char* input = readline("user> ");
    if (!input)
      break;
    add_history(input);

    PRINT(EVAL(READ(input)));

    free(input);
  }

  return EXIT_SUCCESS;
}

char* READ(char* token) {
  return token;
}

char* EVAL(char* expre) {
  return expre;
}

char* PRINT(char* result) {
  printf("=> %s\n", result);
  return result;
}

File: impls/ruby/step1_read_print.rb
require "readline"
require_relative "reader"
require_relative "printer"

def READ(str)
  return read_str(str)
end

def EVAL(expre)
  return expre
end

def PRINT(result)
  print "=> "
  pr_str(result)
end

while buf = Readline.readline("user> ", true)
  PRINT(EVAL(READ(buf)))
end
